Better handling of data corruption

Venemo opened the pull request:
This PR is a follow-up to the data protection PR which added the `DataProtector` class to the secrets daemon. This one deals with what happens when data corruption happens.

* Separate the cases when there is data corruption from when the data is otherwise inaccessible.
* When the data is irretrieveable, delete the data.
* When the data is irretrieveable, send a notification to the user.

Further idea:

* When just the lock code data is gone, maybe prompt the user to set up a new lock code. This is nice to have, but considering how unlikely it is that this actually happens, I have not implemented it yet. Our time is better spent elsewhere.

dcaliste commented on daemon/SecretsImpl/secrets.cpp:
```diff
@@ -2586,3 +2594,23 @@ void Daemon::ApiImpl::SecretsRequestQueue::handleFinishedRequest(
     }
 }
 
+void Daemon::ApiImpl::SecretsRequestQueue::dealWithData() const
```
Didn't try to compile, but why the method is called dealWithData() here, while everywhere you call it with dealWithDataCorruption() (which sounds a better name).

Venemo replied:
Just a typo, will fix.

dcaliste commented on daemon/SecretsImpl/secrets.cpp:
```diff
@@ -2586,3 +2594,23 @@ void Daemon::ApiImpl::SecretsRequestQueue::handleFinishedRequest(
+void Daemon::ApiImpl::SecretsRequestQueue::dealWithData() const
+{
+    // Find secrets directory
+    const static QString systemDataDirPath(QStandardPaths::writableLocation(QStandardPaths::GenericDataLocation) + "/system/");
```
I think this directory is built somewhere else in the code (like the place where to create the databases). It would be better to have a method that return this path, so in case of a change, we need to change only once and don't have to grep in the code to find all the places to change.

dcaliste commented on daemon/SecretsImpl/secrets.cpp:
```diff
@@ -973,13 +978,16 @@ QByteArray Daemon::ApiImpl::SecretsRequestQueue::saltData() const
+        dealWithDataCorruption();
+    } else if (s != DataProtector::Success) {
+        qCWarning(lcSailfishSecretsDaemon) << "saltData: can't read salt data. DataProtector returned:" << s;
+        // TODO: what should we do here? The data is not corrupted, but we can't access it.
```
Maybe we can notify the user here also, like for the data corruption case. In that case, the dealWithDataSomething() can be generalized to notify always and remove the directories only when data is corrupted.

Venemo replied:
Maybe... I'll think about this some more.

chriadam commented:
First step is I think to ask Martin what the appropriate way to handle these cases are, from the user's POV.  E.g. how we should notify them, what sort of UI flow we provide to allow them to attempt to rectify the issue (e.g. re-enter master lock code to regenerate the test cipher data if that one got corrupted) etc.

For the salt data case, I agree that a corruption there is non-recoverable, so deleting the directory is probably the best option.  But I wonder if we should "wait" until the user has been informed and "confirms" the deletion action, instead of doing it immediately.  E.g. maybe in the future we will allow the salt data to be backed up in device backups, so they could then recover the salt data from such a backup, in order to continue using their existing secrets data.

Venemo commented:
@chriadam After talking with @rainemak about this, we came to the conclusion that this is a very very very rare occourence. So (for now at least) my suggestion is this:

* If we are afraid that the data corruption can actually happen, I suggest increasing the number of backup files from 3 to 5.
* Otherwise this PR is just about making sure that the user can continue to use their device without manually needing to drop into the terminal and deleting the corrupted secrets data.
* Since (ideally) this is something end user should never see, we shouldn't spend too much time on polishing it.

All that being said, I don't disagree with polishing this even further in the future, but considering the rarity of the addressed situation my humble opinion is that our time and resources could be better spent elsewhere.

chriadam commented:
Why is a separate request required?  Upon startup, the daemon will know whether the data is corrupted or not, and can store that internally.  Why not simply add an error code which uniquely identifies the issue, and return that for all requests?

chriadam commented on daemon/SecretsImpl/secrets.cpp:
```diff
@@ -2586,3 +2682,54 @@ void Daemon::ApiImpl::SecretsRequestQueue::handleFinishedRequest(
     }
 }
 
+void Daemon::ApiImpl::SecretsRequestQueue::dealWithDataCorruption() const
```
Perhaps not necessary in the first implementation, but in the future it would be good if we dealt with salt data corruption differently to corruption of the test cipher data.  (In the latter case, we can recover by asking the user to re-enter the master lock code and regenerating the test cipher data.)  Perhaps add a parameter which describes the type of corruption (but is currently unused) would be useful, or add a comment regarding the possibility.

Venemo replied:
Honestly, I believe the occourence of this code ever running is so rare that I wouldn't bother with this. But added a comment now about the possibility of doing it in the future.

I would say if a client requests further work on it, let's do this. Otherwise let's not.

chriadam commented on lib/Secrets/healthcheckrequest.cpp:
```diff
@@ -0,0 +1,176 @@
+
+/*!
+ * \class HealthCheckRequest
+ * \brief Allows a client request information about the well being of secrets data.
```
well-being should be hyphenated

Venemo replied:
Will fix!

chriadam commented on lib/Secrets/healthcheckrequest.cpp:
```diff
@@ -0,0 +1,176 @@
+ * \class HealthCheckRequest
+ * \brief Allows a client request information about the well being of secrets data.
+ *
+ * An example of retrieving information about available plugins follows:
```
This line should be "... about the health of the secrets service data follows:"

Venemo replied:
This is just bad copy-paste, thanks for spotting it.

chriadam commented on lib/Secrets/healthcheckrequest.cpp:
```diff
@@ -0,0 +1,176 @@
+ * // status() will change to Finished when complete
+ * // real clients should not use waitForFinished().
+ * req.waitForFinished();
+ * qDebug() << "salt data health:" << req.
```
unfinished

Venemo replied:
Will fix!

Venemo commented:
> Why is a separate request required?

You are not wrong, I could have chosen another approach. However when I started working on it the current approach seemed simpler and cleaner, that is why I went with it.

Venemo commented on daemon/SecretsImpl/secrets.cpp:
```diff
@@ -2114,6 +2185,31 @@ void Daemon::ApiImpl::SecretsRequestQueue::handleFinishedRequest(
             }
             break;
         }
+        case GetHealthInfoRequest: {
```
Yes, you are right. I wasn't 100% sure of this but was wondering myself if this is necessary.

Venemo replied:
Now removed the code from here, and just added a comment that explains why it isn't necessary there.

Venemo commented on daemon/SecretsImpl/secrets.cpp:
```diff
@@ -1223,6 +1261,39 @@ void Daemon::ApiImpl::SecretsRequestQueue::handlePendingRequest(
             }
             break;
         }
+        case GetHealthInfoRequest: {
```
The reason why it's here is because the static variables it uses are in this file. But if you are OK with duplicating those, then I can move it there.

chriadam replied:
Don't duplicate, just pass the secretsDirPath as a parameter, I think.

One other thing to check: which thread do we "normally" read the salt data etc within?  Is it always the main thread?  I assume so, but please double check this.

Venemo replied:
Done.

chriadam commented:
I'm not sure that it is simpler to increase the complexity of the API by adding an entire new request type, compared to adding two error codes, but perhaps I'm wrong.  How is the new request expected to be used - is every client expected to use that request prior to performing any other request, or is it only used if the daemon returns some specific error code to another type of request?

Are you able to prototype the "health error codes rather than health request" solution also for comparison?

chriadam commented:
After discussion on IRC I now agree with Venemo that adding a new request type is the correct solution, although for slightly different reasons.  While I think that it is simpler (from an API semantics perspective) to return the appropriate error code when required, instead of returning the more generic DatabaseLockedError and then requiring the client to perform the HealthCheckRequest to determine the precise issue, one very strong argument in favour of the HealthCheckRequest is that in the future we may need to extend the check to other things (checking individual plugins' metadata databases, for example) and thus the ability to extend the API and information reporting may be very useful.

Venemo commented:
@chriadam I believe I've addressed most of your review findings now. Please go through it one more time and let me know if it's good now. Thank you :)

Venemo commented:
Added support for `HealthCheckRequest` to the tool. Now you can use:

```
secrets-tool --health-check
```

Using this I was able to test the feature and eliminate a couple of bugs in the code.

chriadam commented on daemon/SecretsImpl/secrets.cpp:
```diff
@@ -863,10 +897,12 @@ bool Daemon::ApiImpl::SecretsRequestQueue::determineTestCipherPlugin(
-        qCWarning(lcSailfishSecretsDaemon) << "NOTE: If you are running a pre-release sailfish-secrets version, you need to delete all old data before proceeding";
-        // TODO: find an appropriate solution for dealing with data corruption.
+    if (s == DataProtector::Irretrievable) {
+        qCWarning(lcSailfishSecretsDaemon) << "determineTestCipherPlugin: lock code data is irretrieveably corrupted.";
```
typo, should be irretrievably

chriadam commented on lib/Secrets/healthcheckrequest.cpp:
```diff
@@ -0,0 +1,176 @@
+ * \class HealthCheckRequest
+ * \brief Allows a client request information about the well-being of secrets data.
+ *
+ * An example of retrieving information about the secrets data health follows:
```
Please add some more description to make it clear that this request is really intended for use by the system settings application only, and that most clients will never need to use this request.

chriadam commented on lib/Secrets/secretmanager.cpp:
```diff
@@ -110,6 +110,25 @@ SecretManagerPrivate::getPluginInfo()
+SecretManagerPrivate::getHealthInfo()
+{
+    if (!m_interface) {
+        return QDBusPendingReply<Result>(
```
this needs to return `QDBusPendingReply<Result, Health, Health>`.  I notice that the one above is also incorrect - can you please fix that one also?

chriadam commented:
LGTM, thanks
